stl/inc/chrono

```diff
@@ -2209,6 +2221,2949 @@ namespace chrono {
+                    return _Clock_cast_strategy::_Via_sys_from_utc;
+                } else {
+                    return _Clock_cast_strategy::_None;
+                }
+            }
+        }
+    }
+
+    template <class _DestClock, class _SourceClock, class _Duration>
+    inline constexpr auto _Clock_cast_choice = _Choose_clock_cast<_DestClock, _SourceClock, _Duration>();
+
+    // FUNCTION TEMPLATE clock_cast
+    template <class _DestClock, class _SourceClock, class _Duration,
+        enable_if_t<_Clock_cast_choice<_DestClock, _SourceClock, _Duration> != _Clock_cast_strategy::_None, int> = 0>
+    _NODISCARD auto clock_cast(const time_point<_SourceClock, _Duration>& _Time) {
+        constexpr auto _Strat = _Clock_cast_choice<_DestClock, _SourceClock, _Duration>;
+
+        if constexpr (_Strat == _Clock_cast_strategy::_Direct) {
+            return clock_time_conversion<_DestClock, _SourceClock>{}(_Time);
+        } else if constexpr (_Strat == _Clock_cast_strategy::_Via_sys) {
+            return clock_time_conversion<_DestClock, system_clock>{}(
+                clock_time_conversion<system_clock, _SourceClock>{}(_Time));
+        } else if constexpr (_Strat == _Clock_cast_strategy::_Via_utc) {
+            return clock_time_conversion<_DestClock, utc_clock>{}(
+                clock_time_conversion<utc_clock, _SourceClock>{}(_Time));
+        } else if constexpr (_Strat == _Clock_cast_strategy::_Via_utc_from_sys) {
+            return clock_time_conversion<_DestClock, utc_clock>{}( //
+                clock_time_conversion<utc_clock, system_clock>{}(
+                    clock_time_conversion<system_clock, _SourceClock>{}(_Time)));
+        } else if constexpr (_Strat == _Clock_cast_strategy::_Via_sys_from_utc) {
+            return clock_time_conversion<_DestClock, system_clock>{}( //
+                clock_time_conversion<system_clock, utc_clock>{}(
+                    clock_time_conversion<utc_clock, _SourceClock>{}(_Time)));
+        } else if constexpr (_Strat == _Clock_cast_strategy::_Two_step_ambiguous) {
+            static_assert(_Always_false<_Duration>,
+                "A two-step clock time conversion is required to be unique, "
+                "either through utc_clock or system_clock, but not both (N4878 [time.clock.cast.fn]/2).");
+        } else if constexpr (_Strat == _Clock_cast_strategy::_Three_step_ambiguous) {
+            static_assert(_Always_false<_Duration>,
+                "A three-step clock time conversion is required to be unique, "
+                "either utc-to-system or system-to-utc, but not both (N4878 [time.clock.cast.fn]/2).");
+        } else {
+            static_assert(_Always_false<_Duration>, "should be unreachable");
+        }
+    }
+
+    // [time.parse]
+    struct _Time_parse_fields {
+        using _SubsecondType = duration<int64_t, atto>;
+
+        // These are the primary fields, used to set the chrono type being parsed.
+        optional<int64_t> _Subsecond;

```

Comment:
I am wondering whether a full optional is really needed here. That said I am not crazy enough to do the work, so please disregard

Reply:
I think the main impacts are throughput (but `<chrono>` already instantiates a ton of stuff) and the layout here (as each `optional` spends the equivalent of a `bool` to track `has_value`, but this is mixed with elements that need 4-byte and 8-byte alignment, so there will be lots of padding). It could be possible to improve throughput and save a bit of space, but yes, that would be a *major* refactoring, which is out of scope at this time (we need to complete C++20 *soon*).

The code using `optional` is quite elegant so I am more than happy with the current solution, unless and until we receive feedback that there's a throughput or performance issue.

Reply:
yeah, It really works nice. I blame @CaseyCarter for spoiling me with all those throughput fears

Reply:
I originally did write it with bare integers and `_Invalid_time_field` as a sentinel value, but there are *so many* checks to see if a field is present, it was quite verbose. That's how, as STL noticed, the `_Is_complete` function became a bunch of short `if` statements before he tightened it up. The condition expressions used to be much longer.